pkg/api/pool_test.go

```diff
@@ -0,0 +1,192 @@
+			Kind:   "",
+			Impl:   name,
+			Params: new(structpb.Struct),
+		},
+	}
+)
+
+func TestNewPoolPanics(t *testing.T) {
+	sizeInt := -1 // this has type int; golint complains if we use a descriptive var declaration
+	defer func() {
+		if r := recover(); r == nil {
+			t.Error("Expected value in panic, got nothing.")
+		}
+	}()
+
+	newPool(uint(sizeInt))
+	t.Error("Expected panic in constructor due to size overflow")
+}
+
+func TestPoolSize(t *testing.T) {
+	blockChan := make(chan struct{})
+	testMngr := newTestManager(name, func() (*aspect.Output, error) {
+		<-blockChan
+		return &aspect.Output{Code: code.Code_OK}, nil
+	})
+	b := StaticBinding{
+		RegisterFn: func(r adapter.Registrar) error { return r.RegisterDenyChecker(testMngr) },
+		Manager:    testMngr,
+		Config:     cfg,
+		Methods:    []Method{Check},
+	}
+
+	// Easier than creating a new manager directly and having to register everything. We need all the config either way.
+	mgr, _ := NewMethodHandlers(1, b).(*methodHandlers)
+	underTest := newPool(1)
+
+	// Note: we allocate a result buffer of size two by passing in numAdapters = 2
+	res, enqueue := underTest.requestGroup(mgr.mngr, nil, nil, 2)
+
+	// Enqueue work which will not complete until blockChan is closed; since the pool size == 1 this blocks the queue
+	enqueue(context.Background(), cfg)
+
+	second := make(chan struct{})
+	go func() {
+		// this second enqueue will block until blockChan is closed
+		enqueue(context.Background(), cfg)
+		close(second)
+	}()
+
+	if len(res) != 0 {
+		t.Errorf("Expected result chan to be empty before work completes, got length: %d", len(res))
+	}
+
+	close(blockChan) // unblock the queue
+	<-second         // block the test thread till the second enqueue completes
+
+	// It takes a little time for the two goroutines to write their results; we loop to make the test more reliable.
+	for count := 0; len(res) != 2 && count < 5; count++ {
+		time.Sleep(10 * time.Millisecond)
+	}
+	if len(res) != 2 {
+		t.Errorf("After unblocking queue expected two finished tasks, got %d", len(res))
+	}
+
+	for i := 0; i < 2; i++ {
+		r := <-res
+		if r.out.Code != code.Code_OK {
+			t.Error("Got back bad code.")
+		}
+	}
+}
+
+func TestQuit(t *testing.T) {
+	fail := make(chan struct{})
+	succeed := make(chan struct{})
+	p := newPool(1)
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		close(fail)
+	}()
+
+	go func() {
+		p.shutdown()
+		close(succeed)
+	}()
+
+	select {
+	case <-fail:
+		t.Error("pool.shutown() didn't complete in the expected time")
+	case <-succeed:
+	}
+}
+
+func TestEnqueuePanics(t *testing.T) {
+	p := newPool(1)
+
+	mgr := adapterManager.NewManager(nil)
+
+	numCalls := 1
+	_, enqueue := p.requestGroup(mgr, nil, nil, numCalls)
+	for i := 0; i < numCalls; i++ {
+		enqueue(nil, cfg)
+	}
+
+	defer func() {
+		if r := recover(); r == nil {
+			t.Error("enqueue(nil, nil) got nil, want panic and non-nil recover.")
+		}
+	}()
+	enqueue(nil, cfg)
+	t.Error("enqueue(nil, nil) should panic")
```

Comment:
this isn't `enqueue(nil, nil)` though, is it? cfg looks to be assigned a non-nil value. have i missed something?